File: LeetCode75_LevelTwo/Symmetric_Tree/Symmetric_Tree.cc
// 101. Symmetric Tree C++

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        return checkNodes(root->left, root->right);
    }
    bool checkNodes(TreeNode* left, TreeNode*right) {
        if (left && !right) return false;
        if (!left && right) return false;
        if (!left && !right) return true;
        if (left->val != right->val) return false;
        return checkNodes(left->left, right->right) && checkNodes(left->right, right->left);
    }
};

File: LeetCode75_LevelTwo/Symmetric_Tree/Symmetric_Tree.py
# 101. Symmetric Tree Python
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root: return True
        return self.checkNodes(root.left, root.right)
    def checkNodes(self, left, right):
        if not left and not right: return True
        if not left: return False
        if not right: return False
        if left.val != right.val: return False
        return self.checkNodes(left.left, right.right) and self.checkNodes(left.right, right.left)
